### DISCUSSION TITLE
Default values on record creation?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#523] [Q&A] [answered]
I have a collection where `uid` is the `User` type and is required.

I want to create a record where the `uid` field is always `@request.user.id` and is never allowed to be anything else, even if the client sends `uid` value explicitly.

```ts
client.records.create('myCollection', 
   { uid: client.authState.user.id }   // I don't want to need this or have to trust it
) 
```

**Comment by ganigeorgiev** [comment, marked as answer]
PocketBase doesn't support default values because it is ambigious when they will be applied since we don't have `null` values (you could check this other issues and discussions - https://github.com/pocketbase/pocketbase/issues/42, https://github.com/pocketbase/pocketbase/issues/206).

In general you have 2 options:

1. Restrict the `uid` field with a collection create/update API rule:
    ```
    @request.user.id != "" && uid = @request.user.id
    ```
    Which translates to: "allow submission only if the `uid` (newly submitted for create and previously submitted for update) value match the current authenticated user id"

2. Use a `OnRecordBeforeCreateRequest` hook:
    ```go
    app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
        if e.Record.Collection().Name != "myCollection" {
            return nil // not a "myCollection" record
        }

        // get the authenticated user
        user, _ := e.HttpContext.Get("user").(*models.User)
        if user == nil {
            return rest.NewUnauthorizedError("You are not authorized to perform this action.", nil)
        }

        // always assign the uid to the current authenticated user
        e.Record.SetDataValue("uid", user.Id)

        return nil
    })
    ```
    (_you can also add similar listener for the `OnRecordBeforeUpdateRequest` hook if needed_)

**Reply by benallfree** [reply]
Thanks.

Just for clarification, shouldn't the rule be

```
@request.data.uid = @request.user.id
```

I understand that `@request.data.uid` is the incoming value whereas `uid` is the current value if the record exists. Is that right?

```
uid = @request.user.id
```

Would be useful to restrict reads, I think?

**Reply by ganigeorgiev** [reply]
For the create API rule you can use both `@request.data.uid` or directly the model field because the create rule is applied after a dry-submit within a transaction (this will be documented better after the refactoring).

For list, view, update and delete the model fields contains the old/existing stored value.
Depending on your use case, if you don't want to allow other users to change the uid reference of an existing record, the update API rule should be something like this:
```
uid = @request.user.id && @request.user.id = @request.data.uid
```
Which translates to "Allow only the owner of the record to perform update and disallow uid change".

**Reply by benallfree** [reply]
Oh that’s interesting and somewhat unexpected. How would one use the existing value in validation if it is overwritten before validation occurs? For example Firebase has `data` and `newData`

**Reply by ganigeorgiev** [reply]
>  if it is overwritten before validation occurs?

That's only for the create API rule because you don't have an existing record to check against.

For list, view, update and delete API rules, the model fields represents the "old" state and `@request.data.*` represent the newly submited state so you can compare them if you need to.
